diff --git a/driver/src/egl/egl_platform_android.cpp b/driver/src/egl/egl_platform_android.cpp
index 056e68d..877f431 100644
--- a/driver/src/egl/egl_platform_android.cpp
+++ b/driver/src/egl/egl_platform_android.cpp
@@ -88,7 +88,7 @@ This must be aligned with NUM_FB_BUFFERS in gralloc_priv.h for Jellybean and old
 NUM_BUFFERS = NUM_FB_BUFFERS - 1
 NUM_BUFFERS was reduced to 2 for low memory device. (Original value is 3)
 */
-#define NUM_BUFFERS 2
+#define NUM_BUFFERS 1
 
 #if EGL_ANDROID_recordable_ENABLE
 #include "egl_recordable_android.h"
