diff --git a/drivers/staging/android/ion/ion.c b/drivers/staging/android/ion/ion.c
index 68c7268..49c410d 100644
--- a/drivers/staging/android/ion/ion.c
+++ b/drivers/staging/android/ion/ion.c
@@ -591,8 +591,9 @@ struct ion_handle *ion_alloc(struct ion_client *client, size_t len,
 	struct ion_buffer *buffer = NULL;
 	struct ion_heap *heap;
 	int ret;
+	static int print = 0;
 
-	pr_debug("%s: len %zu align %zu heap_id_mask %u flags %x\n", __func__,
+	trace_printk("%s: len %zu align %zu heap_id_mask %u flags %x\n", __func__,
 		 len, align, heap_id_mask, flags);
 	/*
 	 * traverse the list of heaps available in this system in priority
@@ -606,10 +607,29 @@ struct ion_handle *ion_alloc(struct ion_client *client, size_t len,
 		return ERR_PTR(-EINVAL);
 
 	down_read(&dev->lock);
+
+	if (print++ == 0) {
+		plist_for_each_entry(heap, &dev->heaps, node) {
+			trace_printk("All CMA heaps: heap name:%s, heap id:%d\n",
+					heap->name, heap->id);
+		}
+	}
+
 	plist_for_each_entry(heap, &dev->heaps, node) {
+		if (len == 2490368 && heap_id_mask == 80) {
+			trace_printk("before check: heap name:%s, heap id:%d\n",
+					heap->name,
+					heap->id);
+		}
 		/* if the caller didn't specify this heap id */
-		if (!((1 << heap->id) & heap_id_mask))
+		if (!((1 << heap->id) & heap_id_mask)) {
+			if (len == 2490368 && heap_id_mask == 80) {
+				trace_printk("after check: heap name:%s, heap id:%d\n",
+						heap->name,
+						heap->id);
+			}
 			continue;
+		}
 		buffer = ion_buffer_create(heap, dev, len, align, flags);
 		if (!IS_ERR(buffer))
 			break;
diff --git a/drivers/staging/android/ion/ion_cma_heap.c b/drivers/staging/android/ion/ion_cma_heap.c
index 220afb1..8fa4ec4 100644
--- a/drivers/staging/android/ion/ion_cma_heap.c
+++ b/drivers/staging/android/ion/ion_cma_heap.c
@@ -104,7 +104,8 @@ static int ion_cma_allocate(struct ion_heap *heap, struct ion_buffer *buffer,
 #endif
 	up(&ion_cma_sem);
 	if (!info->cpu_addr) {
-		dev_err(dev, "Fail to allocate buffer\n");
+		dev_err(dev, "Fail to allocate buffer, ION_FLAG_CACHED:%s\n",
+				(buffer->flags & ION_FLAG_CACHED) ? "yes" : "no");
 		goto err;
 	}
 
diff --git a/drivers/staging/android/ion/rockchip/rockchip_ion.c b/drivers/staging/android/ion/rockchip/rockchip_ion.c
index e4d4f2c..5ff2e1e 100644
--- a/drivers/staging/android/ion/rockchip/rockchip_ion.c
+++ b/drivers/staging/android/ion/rockchip/rockchip_ion.c
@@ -63,14 +63,17 @@ static struct ion_heap_desc ion_heap_meta[] = {
 		.type	= ION_HEAP_TYPE_CARVEOUT,
 		.name	= "carveout-heap",
 	}, {
+		//48M
 		.id	= ION_HEAP_TYPE_DMA,
 		.type	= ION_HEAP_TYPE_DMA,
 		.name	= "cma-heap",
 	}, {
+		//16M
 		.id	= ION_HEAP_TYPE_SECURE,
 		.type	= ION_HEAP_TYPE_DMA,
 		.name	= "secured-rga-heap",
 	}, {
+		//80M
 		.id	= ION_HEAP_TYPE_SECURE2,
 		/*.type	= ION_HEAP_TYPE_SECURE,*/
 		.type	= ION_HEAP_TYPE_DMA,
