diff --git a/arch/x86/kernel/pci-dma.c b/arch/x86/kernel/pci-dma.c
index 664ed92..b85540d 100644
--- a/arch/x86/kernel/pci-dma.c
+++ b/arch/x86/kernel/pci-dma.c
@@ -142,6 +142,10 @@ again:
 	if (flag & __GFP_WAIT)
 		page = dma_alloc_from_contiguous(dev, count, get_order(size));
 
+	/* fallback */
+	if (!page)
+		page = alloc_pages_node(dev_to_node(dev), flag, get_order(size));
+
 	if (!page && !retried && (flag & __GFP_WAIT)) {
 		int tmp_lmk_threshold[6] = {0, 0, 0, 0, 0, 32000};
 
@@ -158,9 +162,6 @@ again:
 		pr_debug("__dma_alloc: we %s after aggressive reclaiming!\n",
 			(page ? "Succeed" : "Still Fail"));
 
-	/* fallback */
-	if (!page)
-		page = alloc_pages_node(dev_to_node(dev), flag, get_order(size));
 	if (!page)
 		return NULL;
 
