diff --git a/drivers/staging/android/ion/ion.c b/drivers/staging/android/ion/ion.c
index 68c7268..3d883bac 100644
--- a/drivers/staging/android/ion/ion.c
+++ b/drivers/staging/android/ion/ion.c
@@ -1485,7 +1485,7 @@ end:
 }
 EXPORT_SYMBOL(ion_import_dma_buf);
 
-static int ion_sync_for_device(struct ion_client *client, int fd)
+int ion_sync_for_device(struct ion_client *client, int fd)
 {
 	struct dma_buf *dmabuf;
 	struct ion_buffer *buffer;
@@ -1508,6 +1508,7 @@ static int ion_sync_for_device(struct ion_client *client, int fd)
 	dma_buf_put(dmabuf);
 	return 0;
 }
+EXPORT_SYMBOL(ion_sync_for_device);
 
 /* fix up the cases where the ioctl direction bits are incorrect */
 static unsigned int ion_ioctl_dir(unsigned int cmd)
diff --git a/drivers/staging/android/ion/ion.h b/drivers/staging/android/ion/ion.h
index bd711af..c0da228 100644
--- a/drivers/staging/android/ion/ion.h
+++ b/drivers/staging/android/ion/ion.h
@@ -212,6 +212,8 @@ int ion_share_dma_buf_fd(struct ion_client *client, struct ion_handle *handle);
  */
 struct ion_handle *ion_import_dma_buf(struct ion_client *client, int fd);
 
+int ion_sync_for_device(struct ion_client *client, int fd);
+
 int ion_map_iommu(struct device *iommu_dev, struct ion_client *client,
 		  struct ion_handle *handle, unsigned long *iova,
 		  unsigned long *size);
