diff --git a/core/combo/TARGET_linux-x86.mk b/core/combo/TARGET_linux-x86.mk
index cd86b22..9d7b451 100644
--- a/core/combo/TARGET_linux-x86.mk
+++ b/core/combo/TARGET_linux-x86.mk
@@ -95,7 +95,22 @@ KERNEL_HEADERS := $(KERNEL_HEADERS_COMMON) $(KERNEL_HEADERS_ARCH)
 
 android_config_h := $(call select-android-config-h,target_linux-x86)
 
+ifneq ($(TARGET_BUILD_VARIANT),user)
+TARGET_GLOBAL_CFLAGS += -fno-omit-frame-pointer
+endif
+
+ifeq ($(SMALL_CODE_SIZE),true)
+TARGET_GLOBAL_CFLAGS += \
+       -Os
+else
 TARGET_GLOBAL_CFLAGS += \
+       -O2 \
+       -finline-functions \
+       -finline-limit=300
+endif
+
+TARGET_GLOBAL_CFLAGS += \
+			-Ulinux \
 			-Wa,--noexecstack \
 			-Werror=format-security \
 			-D_FORTIFY_SOURCE=2 \
@@ -112,15 +127,7 @@ TARGET_GLOBAL_CFLAGS += \
 			-I $(dir $(android_config_h)) \
 			-m32
 
-ifeq ($(SMALL_CODE_SIZE),true)
-TARGET_GLOBAL_CFLAGS += \
-       -Os
-else
-TARGET_GLOBAL_CFLAGS += \
-       -O2 \
-       -finline-functions \
-       -finline-limit=300
-endif
+TARGET_GLOBAL_CFLAGS += -mstackrealign -msse3 -mfpmath=sse
 
 ifeq ($(TARGET_DISABLE_TRIPLE_BUFFERING),true)
     TARGET_GLOBAL_CFLAGS += -DTARGET_DISABLE_TRIPLE_BUFFERING
